File: 백준/Gold/17070. 파이프 옮기기 1/파이프 옮기기 1.cc
#include <bits/stdc++.h>

using namespace std;

int n;

int board[20][20];
bool vis[20][20];
int dx[3] = { 0, 1, 1 };
int dy[3] = { 1, 0, 1 };
int ans;

void dfs(int x, int y, int dir) {

	if (x == n  && y == n) {
		ans++;
		return;
	}

	vis[x][y] = 1;

	for (int i = 0; i < 3; i++) {
		if (dir == 0 && i == 1) continue;
		else if (dir == 1 && i == 0) continue;
		else {
			int nx = x + dx[i];
			int ny = y + dy[i];

			if (nx < 0 || nx > n || ny < 0 || ny > n) continue;

			if (i < 2) {
				if (!board[nx][ny] && !vis[nx][ny]) {
					dfs(nx, ny, i);
					vis[nx][ny] = 0;
				}
			}
			else {
				if (!board[nx][ny] && !board[nx - 1][ny] &&
					!board[nx][ny - 1] && !vis[nx][ny])
					dfs(nx, ny, i);
					vis[nx][ny] = 0;
			}
		}
	}
}
int main(void) {
	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> n;

	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			cin >> board[i][j];
		}
	}
	
	dfs(1, 2, 0);

	cout << ans;
}

File: 백준/Gold/17070. 파이프 옮기기 1/파이프 옮기기 1.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
public class Main {
	
	static int N;
	static int[][] board;
	static boolean[][] vis;
	static int[] dx = { 0, 1, 1 };
	static int[] dy = { 1, 0, 1 };
	static int ans;
	
	public static void main(String[] args) throws Exception {
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	
    	N = Integer.parseInt(br.readLine());
    	
    	board = new int[N + 1][N + 1];
    	vis = new boolean[N + 1][N + 1];
    	
    	for(int i = 1; i <= N; i++) {
    		StringTokenizer st = new StringTokenizer(br.readLine());
    		for(int j = 1; j <= N; j++) {
    			board[i][j]	= Integer.parseInt(st.nextToken());
    		}
    	}
    	
    	dfs(1, 2, 0);
    	System.out.println(ans);
	}	
	
	static void dfs(int x, int y, int dir) {
		if(x == N && y == N) {
			ans++;
			return;
		}
		
		vis[x][y] = true;
		
		
		for(int d = 0; d < 3; d++) {
			if(dir == 0 && d == 1) continue;
			if(dir == 1 && d == 0) continue;
			
			int nx = x + dx[d];
			int ny = y + dy[d];
			
			if(nx < 0 || nx > N || ny < 0 || ny > N) continue;
			
			if(d < 2) {
				if(board[nx][ny] == 0 && !vis[nx][ny]) {
					dfs(nx, ny, d);
					vis[nx][ny] = false;
				}
			}
			else {
				if(board[nx][ny] == 0 && board[nx - 1][ny] == 0 &&
						board[nx][ny - 1] == 0 && !vis[nx][ny]) {
					dfs(nx, ny, d);
					vis[nx][ny] = false;
				}
			}
		}
	}
	
}
